### PR TITLE
[MDB IGNORE] Reformats Access IDs for accessibility and futureproofing

### PR DESCRIPTION
**Pull request by Son-of-Space**
## About The Pull Request

This PR changes all access defines from integers to strings and organizes them by department for ease of use. In the process I changed some of the access code for certain items like the airlock electronics or circuits cards to stop converting from text to numbers and just directly reference the access required.

THIS IS MY FIRST REAL REFACTOR so I will need extensive review. I tested pretty much everything and it SEEMS to work but since I know very little about coding I may have unintentionally broken some things or not used the code in an efficient manner.

FOR INSTANCE since we no longer use integers and use strings instead, req_access_txt is no longer necessary. However, that means that we start with varedited lists on doors which I'm told may have an impact (albeit a very minor one) on memory. I intend to go through at a later date and make req_access_txt store the strings and convert to a list properly (as well as manually edit all varedit objects again) if there is, indeed, an impact.

I've also gone through and finished editing the inconsistent standards for varedited windoors/buttons/etc. (biggest one being lawyer accesses on secposts and such)

I also went through and finished fixing many airlocks that still had varedits on them since I could see them all as edited in vsc. Unfortunately the CI didn't detect if it had two req_one_access 's or multiple of a req_access.

## Why It's Good For The Game

This is something that I've wanted to do for a while now but was so work intensive and tedious that I've dropped until now. The purpose of this is to properly standardize how the accesses work and make it very easy for someone looking at the code to understand what does what (without needed to have a separate table to reference the IDs from). It also future-proofs the ability to add IDs onto a department without having to do a significant amount of extra editing in other files besides the access file. It also makes it easier to audit and tweak how these accesses work as a groundwork (for example, now that engineering is properly categorized within itself, it is easier to separate the generic access of ACCESS_CONSTRUCTION into separate, well-partitioned access requirements.)

## Changelog

:cl:
refactor: Compartmentalized defines for access IDs and converts them from integers to strings
fix: Standardizes some access reqs on varedited objects like windoors and shutter buttons
fix: Some missed doors with access reqs instead of mapping helpers were fixed
/:cl:

### CODE REVIEW

**Comment by Son-of-Space** [comment]
I forgot to mark it as draft : (

File: code/__DEFINES/access.dm [hunk 1]
```diff
@@ -1,148 +1,183 @@
...
+#define ACCESS_CREMATORIUM 123
+#define ACCESS_LIBRARY 124
+#define ACCESS_BAR 125
+#define ACCESS_KITCHEN 126
+#define ACCESS_HYDROPONICS 127
+///ID 128-134 currently unused
+/// SERVICE - JOBS
+///ID 135 reserved for HOP in case it changes
+#define ACCESS_JANITOR 136
+#define ACCESS_LAWYER 137
```
Comment (JohnFulpWillard) [inline code/__DEFINES/access.dm:125]: why are these two separated?

Comment (Son-of-Space) [reply code/__DEFINES/access.dm:125]: The janitor also has access to the trash room, and the lawyer is also specialized to other areas. Technically you could rename ACCESS_LAWYER to ACCESS_LAW_OFFICE and move it down but I've tried to touch the names as little as possible I'm thinking about later on removing court access to make it sec entrance since it's kind of unnecessary. When I do that I will probably change the name and move it down.

**Review by Timberpoes — CHANGES_REQUESTED** [review]
I commented this on the Discord a few days ago, but there is absolutely no reason accesses need to be so closely grouped.

We have like, 2 billion usable integer numbers. You can easily separate accesses by the thousands and sub-seperate by the hundreds.

Command-level accesses could start 1000. Sec could start 2000. Engineering could start 3000. So on, so forth.

This gives massive scope for making numbers organised and maybe even meaningful. For example 21XX numbers could be general sec access, 22XX numbers could be job-specific sec access, etc.

Through that, the defines have an organisation to them and have massive scope for modification and expansion.

You can also remove all the comments like `///ID 65-74 currently unused` which are meaningless and don't need to be in our documentation, and reconsider comments like `/// - - - SECURITY 20-39 - - -` which in that example would cascade down to autodoc onto `#define ACCESS_SECURITY 20`

**Comment by Son-of-Space** [comment]
@Timberpoes 

The reason the accesses are closely grouped like this is for natural quality control. You could have command start at 1000, sec at 2000, even 100 and 200... The issue there is that when new accesses are added, what is stopping someone who wasn't aware of this PR from making a new command access be 170, 175, and completely arbitrary? 

No department should ever have more than 20 different access IDs for the sake of sanity and ease of use (especially considering the HOP and new ID trim), so making those large limits with no particular boundary only invites the system to be messed up again somewhere down the line.

I will remove the extra commenting.

**Comment by Timberpoes** [comment]
> @Timberpoes
> 
> The reason the accesses are closely grouped like this is for natural quality control. You could have command start at 1000, sec at 2000, even 100 and 200... The issue there is that when new accesses are added, what is stopping someone who wasn't aware of this PR from making a new command access be 170, 175, and completely arbitrary?

Code review. What stops someone putting a random unrelated access into one of the very small subsets of numbers you've equally arbitrarily assigned?

> No department should ever have more than 20 different access IDs for the sake of sanity and ease of use (especially considering the HOP and new ID trim), so making those large limits with no particular boundary only invites the system to be messed up again somewhere down the line.

"640K [of RAM] ought to be enough for anybody."

When there's no practical difference between a value of 2 thousand and 2 million, it would be wise to use more of the available space if your goal is to have the file itself nicely organised with easy addition of accesses.

In other words, there's no reason not to spread the numbers out. We're not cramming everything into an 8 bit int. I believe BYOND offers us 24 bits of integer precision, which means integer numbers up to ~16 million.

At some point someone will add or remove an access and leave a gap which could be filled by anything too. That's fine as long as a future coder can come along and easily slot in a new access without having to renumber those below it or pick a non-sequential number entirely.

As discussed on Discord though, I think they could probably be strings too and I'm not actually sure if there would be any real negative impact, with a possible positive impact of readability and not having to deal with numbers at all.

**Comment by Son-of-Space** [comment]
I didn't save my changes before I committed oopsy doopsy

**Comment by Son-of-Space** [comment]
Okay, so I broke the RCD somehow and also the mapping helpers. Tattle is working on a fix for the helpers and I'm trying to figure out what I did wrong with the RCD. If I messed that up I probably messed up a few other things too.

**Comment by Son-of-Space** [comment]
Okay so after a very long and frustrating and maddening conversation with Potato/Lemon it looks like the reason some things are breaking is because req_access_txt was converting numbers into a list to be used by req_access. The reason was twofold - 1. byond couldn't handle having strings in maps. 2. lists are a lot more intensive on memory. However, this also meant that some things in the code were looking at the numbers and converting them into a list or vice versa and so things are messing up and this simple reorganization has turned into a full blown refactor that I barely understood and have mostly bruteforced my way through.

Though there IS a light at the end of the tunnel! It looks like I have only a few more things to do:

Given all of that the best course moving forward, as I could parse out:
[] Make sure nothing else is trying to convert accesses from text into numbers or numbers into text because that's no longer relevant since we're using strings!
[] Fix the mapping helpers to properly apply the list of accesses to req_access
AND that's PROBABLY all that needs to be done for this to work.
What else needs to be done to make it the best it possibly can be?
[]make req_access_txt go from converting numbers into a list to converting strings into a list
[]convert all of the 2,000+ manual req_access varedits on all maps that I have edited I think 7 times now back to req_access_txt and edit the list of strings into a string of strings.

**Review by Maurukas — APPROVED** [review]
This passes CI, compiles locally, maps load up without errors. I'm able to join the round and respawn as various roles - all access works as I expect it to. I'm able to unlock and lock the devices I expect to, including those that do not use helpers. I'm able to create and change permissions on cards with Plexagon and the imprinter.

Everything looks good to me. This is quite ambitious and I expect there to be at least a few issues, but given an hour and change of going through it, I can't find any.

**Comment by Son-of-Space** [comment]
image

File: code/__DEFINES/access.dm [hunk 2]
```diff
@@ -1,148 +1,153 @@
...
+	Access for departmental roles will start with the head and go down in level of succession
+	If we ever get to a point where we have more departmental roles than the five (four) available slots, we should be looking to make some job access more generic
+	Access goes from Command, Security, Engineering, Medical, Supply, Science, Service, Away Missions, Mech Access, Admin, then Antag
+	Please try to make the strings for any new accesses as close to the name of the define as possible
+	If you are going to add an access to the list, make sure to also add it to its respective region further below
+	If you're varediting on the map, it uses the string. If you're editing the object directly, use the define name*/
+
+/// Command general access, EVA storage windoors, gateway shuters, AI integrity restorer, comms console
+#define ACCESS_COMMAND "command"
+#define ACCESS_AI_UPLOAD "ai upload"
```
Comment (JohnFulpWillard) [inline code/__DEFINES/access.dm:13]:
```suggestion
#define ACCESS_AI_UPLOAD "ai_upload"
```
Why do you use spaces here instead of underscores? Makes it easier to mess up varediting if anything.

File: code/modules/modular_computers/file_system/programs/card.dm [hunk 3]
```diff
@@ -59,7 +59,7 @@
 	var/list/managers = SSid_access.sub_department_managers_tgui
 	for(var/access_as_text in managers)
 		var/list/info = managers[access_as_text]
-		var/access = text2num(access_as_text)
+		var/access = (access_as_text)
```
Comment (JohnFulpWillard) [inline code/modules/modular_computers/file_system/programs/card.dm:62]:
```suggestion
		var/access = access_as_text
```

File: code/__DEFINES/access.dm (hunk 2 above)
Comment (Son-of-Space) [reply code/__DEFINES/access.dm:13]: because it isn't a define and it's easier to hit the spacebar than hit the underscore.

Comment (san7890) [reply code/__DEFINES/access.dm]: image

File: code/controllers/subsystem/id_access.dm [hunk 4]
```diff
@@ -248,26 +248,26 @@ SUBSYSTEM_DEF(id_access)
 	desc_by_access["[ACCESS_SECURITY]"] = "Security"
 	desc_by_access["[ACCESS_BRIG]"] = "Holding Cells"
 	desc_by_access["[ACCESS_COURT]"] = "Courtroom"
-	desc_by_access["[ACCESS_FORENSICS]"] = "Forensics"
+	desc_by_access["[ACCESS_DETECTIVE]"] = "Detective"
```
Comment (JohnFulpWillard) [inline code/controllers/subsystem/id_access.dm:251]:
```suggestion
	desc_by_access["[ACCESS_DETECTIVE]"] = "Detective Office"
```

File: _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm [hunk 5]
```diff
@@ -66,11 +66,10 @@
 /obj/structure/cable,
 /obj/effect/decal/cleanable/dirt,
 /obj/machinery/door/airlock/maintenance_hatch{
-	name = "General Storage Access";
-	req_one_access_txt = "12"
+	name = "General Storage Access"
 	},
 /obj/effect/mapping_helpers/airlock/locked,
-/turf/open/floor/catwalk_floor,
+/obj/effect/mapping_helpers/airlock/access/all/engineering/maintenance/turf/open/floor/catwalk_floor,
```
Comment (san7890) [inline _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm:72]:
```suggestion
/obj/effect/mapping_helpers/airlock/access/all/engineering/maintenance,
/turf/open/floor/catwalk_floor,
```

Comment (san7890) [reply _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm:72]: FUCK

Comment (san7890) [reply _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm]: WHO DID THIS

Comment (san7890) [reply _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm]: GRRRRAAAAAA

Comment (san7890) [reply _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm]: FUCK

Comment (san7890) [reply _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm]: YOU MADE ME WEBEDIT

File: code/game/machinery/doors/windowdoor.dm [hunk 6]
```diff
@@ -432,7 +432,7 @@
 
 /obj/machinery/door/window/brigdoor/security/holding
 	name = "holding cell door"
-	req_one_access = list(ACCESS_BRIG_ENTRANCE) //lawyer also gets brig_entrance
+	req_one_access = list(ACCESS_SECURITY)
```
Comment (JohnFulpWillard) [inline code/game/machinery/doors/windowdoor.dm:435]:
```suggestion
	req_one_access = list(ACCESS_BRIG_ENTRANCE)
```
Why remove lawyer access to this?

File: code/game/objects/structures/crates_lockers/closets/secure/security.dm [hunk 7]
```diff
@@ -228,8 +228,8 @@
 /obj/structure/closet/secure_closet/evidence
 	anchored = TRUE
 	name = "Secure Evidence Closet"
-	req_access_txt = "0"
-	req_one_access_txt = list(ACCESS_ARMORY, ACCESS_FORENSICS)
+	req_access = null
```
Comment (JohnFulpWillard) [inline code/game/objects/structures/crates_lockers/closets/secure/security.dm:231]:
```suggestion
```
this isn't needed because its not a subtype of anything that has req_access set

File: _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm (hunk 5 above)
Comment (san7890) [reply _maps/map_files/tramstation/modular_pieces/maintenance_storagebig_2.dmm]: dammit

File: code/game/machinery/doors/windowdoor.dm (hunk 6 above)
Comment (Son-of-Space) [reply code/game/machinery/doors/windowdoor.dm:435]: because it's a holding cell used for mass holding people. Why would a lawyer who doesn't have a mindshield be able to open the door for them to escape? They aren't security.
